internal/utils/ringbuffer/ringbuffer.go

```diff
@@ -0,0 +1,73 @@
+package ringbuffer
+
+type RingBuffer[T any] struct {
+	ring             []T
+	headPos, tailPos int
+	full             bool
+}
+
+func (r *RingBuffer[T]) Init(size int) {
+	r.ring = make([]T, size)
+}
+
+func (r *RingBuffer[T]) Len() int {
+	if r.full {
+		return len(r.ring)
+	}
+	if r.tailPos >= r.headPos {
+		return r.tailPos - r.headPos
+	}
+	return r.tailPos - r.headPos + len(r.ring)
+}
+
+func (r *RingBuffer[T]) PushBack(t T) {
+	if r.full || len(r.ring) == 0 {
+		r.grow()
+	}
+	r.ring[r.tailPos] = t
+	r.tailPos++
+	if r.tailPos == len(r.ring) {
+		r.tailPos = 0
+	}
+	if r.tailPos == r.headPos {
+		r.full = true
+	}
+}
+
+func (r *RingBuffer[T]) PopFront() (t T, ok bool) {
+	if r.Len() == 0 {
```

Comment:
This should be correct, right?
Will probably speed up the benchmark.

```suggestion
	if r.tailPos == r.headPos {
```

Reply:
`r.tailPos == r.headPos` means the queue is full or empty. So it should be
```
if !r.full && r.tailPos == r.headPos
```

Reply:
Right. Probably still more performant than calling `Len`.